experimental/cache_analysis.py

```diff
@@ -0,0 +1,139 @@
+from matplotlib.pyplot import figure
+import matplotlib.pyplot as plt
+import io
+import numpy as np
+import pandas as pd
+from statistics import mean, median
+import collections
+import csv
+import seaborn as sns
+
+""" This functions draws two types of plots for cases where each x value has only one y value.
+
+Here it is used to draw plots for stats (ie mean and median of all runs).
+It can be used for drawing each run individually.
+"""
+def draw_single_run_plot(data, name):
+  figure(num=None, figsize=(10, 8), dpi=80, facecolor='w', edgecolor='k')
+
+  x = list(data.keys())
+  y = list(data.values())
+
+  plt.subplot(211)
+  plt.plot(x,y)
+  plt.subplot(212)
+  plt.scatter(x,y)
+  plt.savefig(name)
+  plt.cla()
+
+
+""" This functions draws boxplot for all data points.
+"""
+def draw_boxplot(data, x, name):
+  fig = plt.figure(figsize=(30, 10))
+  ax = fig.add_subplot(111)
+  boxes = []
+  for l in data:
+    c = 0
+    for i in l:
+      boxes.append([x[c], i])
+      c += 1
+  dd = pd.DataFrame(boxes, columns = ["size (bytes)", "time (cycles)"])
+  sns.set_theme(style="whitegrid")
+
+  ax = sns.boxplot(data = dd, showfliers=False, x = 'size (bytes)', y = 'time (cycles)')
+  ax.grid(False)
+  fig = ax.get_figure()
+  fig.savefig(name)
+
+""" This functions draw all data points in a plot in log scale
+"""
+def draw_all_datapoints(data, sizes, name):
+  fig = plt.figure(figsize=(30, 10))
+  ax = fig.add_subplot(111)
+  for l in data:
+    ax.scatter(sizes,l)
+  # start, end = ax1.get_xlim()
+  # ax1.xaxis.set_ticks(np.arange(0, end))
+  # start, end = ax1.get_ylim()
+  # ax1.yaxis.set_ticks(np.arange(0, end))
+  ax.set_yscale('log')
+  ax.set_xscale('log', base = 2)
+  plt.xlabel("size (bytes)")
+  plt.ylabel("time (cycles)")
+  plt.title(" all data points")
+  plt.grid(False)
+  plt.savefig(name)
+  plt.cla()
+
+
+
+file_name = 'cache_line_size_results.csv'  # name of result file
+file_path = '/home/negara/safeside/build/experimental/'  # path to the result file
```

Comment:
This won't work on my computer :(, and needs a different way to get the file path.

One option is to take the whole path (file name and directory) as an input (e.g. `sys.argv[1]`). Another is to always write to a temporary file, and print the filename so that the user can copy-paste (`with tempfile.NamedTemporaryFile(delete=False) as f: ...` and `print("Output:", f.name)`, see [here](https://docs.python.org/3/library/tempfile#tempfile.NamedTemporaryFile)). Last option is to output to stdout (i.e. use `print`) and let the caller redirect output to a file: `python3 cache_analysis.py > cache_line_size_results.csv`.

I think of these three: the first is the most finicky, and the last would be complex because there's also output for the user (which must now be sent to stderr instead). So I think the best option is to use NamedTemporaryFile.

Reply:
I did not understand how to use the second option for reading a file! 
For now, I changed it to the first option and with a TODO comment.